cou/apps/app.py

```diff
@@ -182,7 +305,251 @@ def _get_os_origin(self) -> str:
         """
         for origin in ("openstack-origin", "source"):
             if self.config.get(origin):
+                self.origin_setting = origin
                 return self.config[origin].get("value", "")
 
         logger.warning("Failed to get origin for %s, no origin config found", self.name)
         return ""
+
+    async def _check_upgrade(self, target: OpenStackRelease) -> None:
+        """Check if an application has upgraded its workload version.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        """
+        status = await async_get_status()
+        app_status = status.applications.get(self.name)
+        units_not_upgraded = []
+        for unit in app_status.units.keys():
+            workload_version = app_status.units[unit].workload_version
+            compatible_os_versions = OpenStackCodenameLookup.lookup(self.charm, workload_version)
+            if target not in compatible_os_versions:
+                units_not_upgraded.append(unit)
+        if units_not_upgraded:
+            logger.error(
+                "App: '%s' has units: '%s' didn't upgrade to %s",
```

Comment:
Unit names already contain the app name so this message could be more compact
```suggestion
                "Units '%s' failed to upgrade to %s",
```